src/axom/core/Map.hpp

```diff
@@ -0,0 +1,487 @@
+   *
+   * \param [in] buckets user-specified number of buckets in new Map. -1 to fall back to multiplicative factor.
+   * \param [in] factor user-specified multiplicative factor to determine size of new Map based upon existing. 
+   *
+   * \note if neither input is specified, the new Map will be of size 2*old Map size. 
+   * \note both map instances exist in memory until rehash returns. Risks running out of memory.
+   */  
+  void rehash(int buckets = -1, int factor = -1){
+   
+    int newlen = 0;
+    IndexType ind;
+    std::size_t hashed;
+    if(buckets != -1 && buckets > m_bucket_count){
+      newlen = buckets;
+    }
+    else if(factor != -1 && factor > 1){
+      newlen = factor*m_bucket_count;
+    }
+    else{
+      newlen = 2*m_bucket_count;
+    }
+
+    axom_map::Bucket<Key,T> * new_list = alloc_map(newlen, m_bucket_len);
+
+    for(int i = 0; i < m_bucket_count; i++){
+      ind = m_buckets[i].m_head;
+      while(ind != -1){
+        hashed = (get_hash(m_buckets[i].m_list[ind].key) % newlen);
+        new_list[hashed].insert_no_update(m_buckets[i].m_list[ind].key, m_buckets[i].m_list[ind].value);     
+        ind = m_buckets[i].m_list[ind].next;
+      }
+      axom::deallocate(m_buckets[i].m_list);
+    }
+    axom::deallocate(m_buckets);
+    m_buckets = new_list;
+    m_bucket_count = newlen;
+    return; 
+  } 
+
+  /*!
+   * \brief Inserts given key-value pair into the Map instance.
+   *
+   * \param [in] key the key used to index into the Map to store this item
+   * \param [in] val the data value of the pair to insert into the map
+   *
+   * \note Deviation from STL unordered_map is that insertion fails when map full, rather than automatically rehashing. 
+   *  Manual call of rehash() required.
+   *
+   * \return A pair whose first element is a pointer to the inserted item and bool set to true
+   *  if successful. Otherwise, the second element is set to false, and the first to one of two values: a sentinel node 
+   *  if the map is overfilled, and the actual node with the given key if an item with given key already exists.
+   *
+   */
+  axom_map::Pair<Key, T> insert(Key key, T val){
+    axom_map::Bucket<Key,T> * target = get_bucket(get_hash(key)); 
+    //Candidate to get cut out if branching becomes too much of an issue.
+    axom_map::Pair<Key, T> ret = target->insert_no_update(key, val);
+    if(ret.second == true){ m_size++; }
+    return ret;
+  }
+
+  /*!
+   * \brief Removes key-value pair with given key from the Map instance.
+   *
+   * \param [in] key the key of the item to be removed from the Map instance.
+   *
+   * \note Deviation from STL unordered_map in returning a boolean instead of an iterator to next item in map.
+   *  Mostly useful on sparing logic when this moves to paralellism, since the bucket boundaries would become inconvenient.
+   *
+   * \return A bool value of true if the item was successfully removed, false otherwise. 
+   */  
+  bool remove(Key key){
```

Comment:
API doesn't return the value, nor iterator to the value, but an iterator to the next member, when it returns an iterator, which is unfortunate at bucket boundaries.

However, now that I look, if I return a size_t, then this actually matches the behavior for erase(Keytype), because it only returns an iterator when you call erase with an iterator. Could this just be switched from a bool to a size_t (or IndexType), to match the 3rd version at that link?